kubernetes-volume: add dimensions for other volume types

pdecat opened the pull request:
Add dimensions to kubernetes-volume monitor for:
- persistent volumes: GCE
- non persistent volumes: ConfigMaps, DownwardAPI, EmptyDir, and Secret

pdecat commented on pkg/monitors/kubernetes/volumes/pvdim.go:
```diff
@@ -64,17 +67,32 @@ func dimsForPersistentVolumeSource(pvs v1.PersistentVolumeSource) map[string]str
+
+	// NON PERSISTENT TYPES:
+	case vs.ConfigMap != nil:
+		return map[string]string{"volume_type": "configMap"}, nil
```
Not 100% about which case to adopt here.

keitwb replied:
I would do `ConfigMap` since that is so prevalent in K8s.

pdecat replied:
Re-reading the docs about the `volume_type` dimension (https://github.com/signalfx/signalfx-agent/blob/v5.3.1/pkg/monitors/kubernetes/volumes/metadata.yaml#L22-L24):

> The type of the underlying volume -- this will be *the key used in the k8s volume config spec* (e.g. awsElasticBlockStore, etc.)

and the mentioned the volume spec reference (https://kubernetes.io/docs/concepts/storage/volumes/#configmap), it turns out `configMap` is probably better after all.

keitwb replied:
Ok, sounds good!

keitwb commented on pkg/monitors/kubernetes/volumes/pvdim.go:
```diff
@@ -47,10 +47,13 @@ func findVolumeInPod(pod *v1.Pod, volName string) *v1.Volume {
 func dimsForPersistentVolumeSource(pvs v1.PersistentVolumeSource) map[string]string {
-	// IF YOU ADD A NEW TYPE HERE, ADD IT IN dimsForVolumeSource BELOW TOO!
+	// IF YOU ADD A NEW PERSISTENT TYPE HERE, ADD IT IN dimsForVolumeSource BELOW TOO!
 	if pvs.AWSElasticBlockStore != nil {
```
Linter is saying that this has enough cases to be a switch.

pdecat replied:
Done.

pdecat commented:
Added the docs to the metadata.yaml file but I'm having a hard time at generating the other files:

```
make SHELL="/bin/bash -vx" docs
bash -c "rm -f docs/{observers,monitors}/*"
bash -c "rm -f docs/{observers,monitors}/*"
+ bash -c 'rm -f docs/{observers,monitors}/*'
scripts/docs/make-docs
scripts/docs/make-docs
+ scripts/docs/make-docs
+++ dirname scripts/docs/make-docs
++ cd scripts/docs
++ pwd
+ SCRIPT_DIR=/home/patrick/go/src/github.com/signalfx/signalfx-agent/scripts/docs
+ . /home/patrick/go/src/github.com/signalfx/signalfx-agent/scripts/docs/../code-gen-helpers.sh
++ AGENT_BIN=signalfx-agent
++++ dirname /home/patrick/go/src/github.com/signalfx/signalfx-agent/scripts/docs/../code-gen-helpers.sh
+++ cd /home/patrick/go/src/github.com/signalfx/signalfx-agent/scripts/docs/..
+++ pwd
++ MY_SCRIPT_DIR=/home/patrick/go/src/github.com/signalfx/signalfx-agent/scripts
++ selfdescribe_json=/home/patrick/go/src/github.com/signalfx/signalfx-agent/scripts/../selfdescribe.json
+++ getconf _NPROCESSORS_ONLN
++ NUM_CORES=12
+++ cat
++ doc_types='{
  "slice": "list",
  "uint16": "integer",
  "uint": "unsigned integer",
  "int": "integer",
  "struct": "object",
  "interface": "any"
}'
+ generate_selfdescribe_json
+ signalfx-agent selfdescribe
[0000]  WARN sendAll was specified on monitor type 'collectd/marathon' but some metrics were also marked as 'default'
[0000]  WARN Could not unmarshal default value `{{.host}}-{{.port}}` for field BrokerName error=invalid character '{' looking for beginning of object key string
[0000]  WARN sendAll was specified on monitor type 'internal-metrics' but some metrics were also marked as 'default'
+ echo ''
+ inject_to_obj doc_types '{
  "slice": "list",
  "uint16": "integer",
  "uint": "unsigned integer",
  "int": "integer",
  "struct": "object",
  "interface": "any"
}'
+ jq -r '. + {doc_types: {
  "slice": "list",
  "uint16": "integer",
  "uint": "unsigned integer",
  "int": "integer",
  "struct": "object",
  "interface": "any"
}}'
+ j .
+ jq -r .
+ gomplate --file /home/patrick/go/src/github.com/signalfx/signalfx-agent/scripts/docs/templates/monitor-main.md.tmpl --datasource agent=stdin:///agent.json

+ j .
+ jq -r .
+ inject_to_obj doc_types '{
  "slice": "list",
  "uint16": "integer",
  "uint": "unsigned integer",
  "int": "integer",
  "struct": "object",
  "interface": "any"
}'
+ gomplate --file /home/patrick/go/src/github.com/signalfx/signalfx-agent/scripts/docs/templates/observer-main.md.tmpl --datasource agent=stdin:///agent.json
+ jq -r '. + {doc_types: {
  "slice": "list",
  "uint16": "integer",
  "uint": "unsigned integer",
  "int": "integer",
  "struct": "object",
  "interface": "any"
}}'

+ j .
+ inject_to_obj doc_types '{
  "slice": "list",
  "uint16": "integer",
  "uint": "unsigned integer",
  "int": "integer",
  "struct": "object",
  "interface": "any"
+ jq -r .
}'
+ jq -r '. + {doc_types: {
  "slice": "list",
  "uint16": "integer",
  "uint": "unsigned integer",
  "int": "integer",
  "struct": "object",
  "interface": "any"
}}'
+ gomplate --file /home/patrick/go/src/github.com/signalfx/signalfx-agent/scripts/docs/templates/top-config.md.tmpl --datasource agent=stdin:///agent.json

22:26:53 FTL  error="template: /home/patrick/go/src/github.com/signalfx/signalfx-agent/scripts/docs/templates/top-config.md.tmpl:9:249: executing \"configTable\" at <(eq .elementKind \"struct\") and (ne .elementKind \"interface\")>: can't give argument to non-function eq .elementKind \"struct\""
make: *** [Makefile:174: docs] Error 1

```

Note: I modified the make-docs scripts to get more details:

```
git diff scripts/docs/make-docs
diff --git scripts/docs/make-docs scripts/docs/make-docs
index 0619a538..cc41fd15 100755
--- scripts/docs/make-docs
+++ scripts/docs/make-docs
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-set -euo pipefail
+set -euox pipefail
 
 SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```

pdecat commented:
Alright, building gomplate with go 1.13.8 instead of 1.14.4 resolved the above issue.

Now `make docs` generated `docs/monitors/kubernetes-volumes.md` properly but:

- there some warning and error messages:
  ```
  make docs
  bash -c "rm -f docs/{observers,monitors}/*"
  scripts/docs/make-docs
  [0000]  WARN sendAll was specified on monitor type 'collectd/marathon' but some metrics were also marked as 'default'
  [0000]  WARN Could not unmarshal default value `{{.host}}-{{.port}}` for field BrokerName error=invalid character '{' looking for beginning of object key string
  [0000]  WARN sendAll was specified on monitor type 'internal-metrics' but some metrics were also marked as 'default'
  [0000] ERROR Found metadata for kubelet-metrics monitor in pkg/monitors/kubernetes/kubeletmetrics/metadata.yaml but it doesn't appear to be registered
  [0000] ERROR Found metadata for mongodb-atlas monitor in pkg/monitors/mongodb/atlas/metadata.yaml but it doesn't appear to be registered
   ```

- other files were also re-generated:
    ```
    git status
    On branch kubernetes_more_volume_types
    Your branch is up to date with 'origin/kubernetes_more_volume_types'.

    Changes to be committed:
    (use "git restore --staged <file>..." to unstage)
            modified:   docs/monitors/kubernetes-volumes.md

    Changes not staged for commit:
    (use "git add/rm <file>..." to update what will be committed)
    (use "git restore <file>..." to discard changes in working directory)
            modified:   docs/config-schema.md
            modified:   docs/monitor-config.md
            modified:   docs/monitors/collectd-php-fpm.md
            modified:   docs/monitors/collectd-rabbitmq.md
            deleted:    docs/monitors/kubelet-metrics.md
            deleted:    docs/monitors/mongodb-atlas.md
            modified:   docs/monitors/supervisor.md
            modified:   docs/monitors/telegraf-dns.md
            modified:   selfdescribe.json
    ```

keitwb commented:
Yeah sorry the doc generation process is dependent on gomplate 3.4.0 (I install straight from their precompiled binary on github instead of compiling from source).  I think there were some subtle changes to the go template library in 1.14 that changed how it rendered nils or something.   I'll try and streamline the install/run process for the docs for future contributors.

Anyway, I commited a small commit to update the selfdescribe.json file that somehow didn't get updated.  I'll merge one tests pass.

Thanks again for adding this.

pdecat commented:
Thanks @keitwb! 🎉

pdecat commented:
FTR, it is working great!

![image](https://user-images.githubusercontent.com/318490/84984800-dbf9f600-b13b-11ea-8c8c-b9d601393d11.png)
